### PR TITLE
[POAE7-2648] support null in filter and expr

### PR DESCRIPTION
**Pull request by Ma-Jian1**
### What changes were proposed in this pull request?
support null in filter and expr
disable implicit cast for bool for some operator overload.

### Why are the changes needed?
next-gen compiler

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
UT

### Which label does this PR belong to?
INFRA

### CODE REVIEW

File: cider/type/plan/ConstantExpr.cpp [hunk 1]
```diff
@@ -34,7 +34,8 @@ JITExprValue& Constant::codegen(JITFunction& func) {
       CIDER_THROW(CiderCompileException,
                   "NULL type literals are not currently supported in this context.");
     case kBOOLEAN:
-      return set_expr_value(func.createConstant(getJITTag(type), get_constval().boolval));
+      return set_expr_value(nullptr,
```
Comment (yma11) [inline cider/type/plan/ConstantExpr.cpp:37]: `nullptr` should be `func.createConstant(getJITTag(type), false)` to indicate it's a not null constant.

File: cider/type/plan/BinaryExpr.cpp [hunk 2]
```diff
@@ -86,41 +79,45 @@ JITExprValue& BinOper::codegen(JITFunction& func) {
...
 
-JITExprValue& BinOper::codegenFixedSizeColCmpFun(JITValue& lhs, JITValue& rhs) {
+JITExprValue& BinOper::codegenFixedSizeColCmpFun(JITValuePointer& null,
+                                                 JITValue& lhs,
+                                                 JITValue& rhs) {
   // TODO: Null Process
   switch (get_optype()) {
     case kEQ:
-      return set_expr_value(nullptr, lhs == rhs);
+      return set_expr_value(null, lhs == rhs);
```
Comment (yma11) [inline cider/type/plan/BinaryExpr.cpp:110]: maybe we should refine this by adding a new API `set_expr_null` to avoid passing the null everywhere?

Comment (yma11) [reply cider/type/plan/BinaryExpr.cpp:110]: maybe you should use `FixSizeJITExprValue`.

File: cider/type/plan/ConstantExpr.cpp (hunk 1 above)
Comment (Ma-Jian1) [reply cider/type/plan/ConstantExpr.cpp:37]: It's ok to create a false constant, but then we will need to check null in IR. any suggestions? @bigPYJ1151

File: cider/type/plan/BinaryExpr.cpp (hunk 2 above)
Comment (Ma-Jian1) [reply cider/type/plan/BinaryExpr.cpp:110]: It's just null value, should be passed whether it is nullptr or not.

File: cider/type/plan/ConstantExpr.cpp (hunk 1 above)
Comment (bigPYJ1151) [reply cider/type/plan/ConstantExpr.cpp:37]: I think the nullability of Constant should always be false except Constant with kNULLT type. So we could pass a nullptr in most cases.

Comment (yma11) [reply cider/type/plan/ConstantExpr.cpp:37]: If we set the null to `nullptr`, it probably cause problem when trying to use `expr.getNull()` when we don't know its a constant or not.

Comment (Ma-Jian1) [reply cider/type/plan/ConstantExpr.cpp:37]: I add an API `isNullable()`, we should always check it

**Review by yma11 — APPROVED** [review]

**Pull request merged**